doc/key-synchronization.md

```diff
@@ -0,0 +1,145 @@
+# Enclave key synchronization
+
+Nitriding supports horizontal scaling, i.e., it allows for the synchronization
+of key material among identical enclaves.  Key material consists of both
+_nitriding_ and _application_ keys:
+
+1. Nitriding's key material is the self-signed HTTPS certificate (both public
+   and private key) that provides the confidential channel between clients and
+   the enclave.
+2. The application's key material is application-specific.  Nitriding is
+   agnostic to the structure of this key material and treats it as arbitrary
+   bytes.
+
+All of the above must be synced among enclaves.
+
+For enclave key synchronization to work, there must be a _single leader
+enclave_ and _one or more worker enclaves_.  The leader's sole job is to
+create key material and make itself available for synchronizing this key
+material with worker enclaves.  Worker enclaves do the actual work, i.e.,
+process user requests.  Before doing any work though, workers must register
+themselves with the leader, which triggers key synchronization.
+
+To set up key synchronization, several steps are necessary:
+
+* Use the `-fqdn-leader` command line flag on both the leader and the worker.
+  Note that the leader and worker images _must be identical_.  The leader is
+  only willing to synchronize key material with _identical enclaves_.
+* Practically speaking, the leader is meant to run in a separate k8s deployment
+  from the workers.
+
+## Protocol
+
+1. The leader creates a 20-byte nonce $\textrm{nonce}_l$ and sends it to the
+   worker as part of a `GET` request.
+2. Upon receiving $\textrm{nonce}_l$, the worker creates its own 20-byte nonce
+   $\textrm{nonce}_w$ and an ephemeral asymmetric key pair $K_e = \(sk, pk\)$,
+   which we generate with Go's `crypto/nacl/box` package. The worker now asks
+   its hypervisor to create an attestation document $A_w$ containing
```

Comment:
This attestation document from the worker contains the image ID right? I think that's what is being described in step 3. The check to make sure the image ID matches establishes a known method of resolving the trust on first use problem that would be established if the worker only sent an ephemeral key.

Reply:
Yes, attestation documents always contain the image ID. I added a clarifying sentence in e36b808.